### PR TITLE
build: bump DaCe package version

### PR DESCRIPTION
**Pull request by edopao**
Upgrade DaCe package version to release v0.16.1. Include GT4Py-next code changes enabled by bugfixes included in the new DaCe package.

Package versions updated in requirement files, but keeping `nanobind==1.9.2` (GT4Py does not support v2.0.0) and `mypy==1.10.0` (v1.10.1 not available for pre-commit).

A couple of code changes in GT4Py-cartesian DaCe backend:
- temporary config in test case to force serialize all fields in SDFG JSON representation, because the default config value was chaged in the new dace release
- handle scalar parameters in stencil objects as scalars rather than symbols, to avoid undefined symbol issues when the parameter is not used

### CODE REVIEW

**Review by egparedes — COMMENTED** [review]
Have you run the tox action (`tox r -e requirements-base`) to automatically update the requirements (see: https://github.com/GridTools/gt4py/blob/main/docs/development/tools/requirements.md)?

**Comment by edopao** [comment]
> Please run the tox action (`tox r -e requirements-base`) to automatically update all the requirement files from the project metadata (see: https://github.com/GridTools/gt4py/blob/main/docs/development/tools/requirements.md)

I was waiting to do this step for the fix on `nanobind` version. Anyway, I have updated the requirement files now but I manually force `nanobind==1.9.2`

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -66,7 +66,7 @@ repos:
   ## version = re.search('mypy==([0-9\.]*)', open("constraints.txt").read())[1]
   ## print(f"#========= FROM constraints.txt: v{version} =========")
   ##]]]
-  #========= FROM constraints.txt: v1.10.0 =========
+  #========= FROM constraints.txt: v1.10.1 =========
   ##[[[end]]]
   rev: v1.10.0  # MUST match version ^^^^ in constraints.txt (if the mirror is up-to-date)
```
Comment (egparedes) [inline .pre-commit-config.yaml:71]:
```suggestion
  rev: v1.10.1  # MUST match version ^^^^ in constraints.txt (if the mirror is up-to-date)
```

Comment (edopao) [reply .pre-commit-config.yaml:71]: It seems that this version is not available for pre-commit:
```
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy.
An unexpected error has occurred: CalledProcessError: command: ('/usr/bin/git', 'checkout', 'v1.10.1')
return code: 1
stdout: (none)
stderr:
    error: pathspec 'v1.10.1' did not match any file(s) known to git
```

**Review by egparedes — APPROVED** [review]
LGTM

**Comment by edopao** [comment]
@FlorianDeconinck Do you need to run any test on PACE before I merge this PR?

**Review by FlorianDeconinck — APPROVED** [review]
LGTM

**Comment by FlorianDeconinck** [comment]
> @FlorianDeconinck Do you need to run any test on PACE before I merge this PR?

We will, but it shouldn't block this. We have commit-pinned versions of both gt4py and dace, we will move them with full testing on our schedule and report if anything breaks

**Comment by FlorianDeconinck** [comment]
Btw there's another bug in Cartesian with unused scalar we found, which this PR doesn't solve. Frank Malentino from NOAA has the specifics if anyone is interested.

**Pull request merged**